Document using BaseTableCoordinate to make a gWCS (#414)
sunpy/ndcube: docs/tabular_coordinates.rst

.. _tabular_coordinates:

===================
Tabular Coordinates
===================

> The documentation so far has assumed that you are constructing an NDCube object from a WCS object.
> Either one that you have read from a file or explicitly created.
> In this section we will discuss the functionality ndcube provides for constructing WCSes from tables of coordinate values.

Reviewer: ```suggestion
So far we have assumed that we are constructing an NDCube using a WCS that has been read from a file or explicitly created.
However, in some cases we may have tables giving the coordinate values at each pixel and converting these into a WCS manually can be tedious.
Therefore ndcube provides tools for constructing WCSes from such tables.
```

Tabular coordinates are useful when there is no mathematical description of the axis, or when it's a natural fit for the data you have.

> It's worth considering that tabular coordinates are generally not as polished as a functional transform in a WCS, if you can build a functional WCS for your coordinate system then that is highly recommended.

Reviewer: ```suggestion
It's worth considering that tabular coordinates are generally not as polished as a functional transform in a WCS.
Therefore, if possiible, building a functional WCS for your coordinate system is highly recommended.
```

Tabular Coordinates and WCSes
=============================

All coordinate information in ndcube is represented as a WCS.

> Even the `.ExtraCoords` class which allows the user to add tabular data to axes generates a WCS which is then used by the coordinate functions and plotting in ndcube.
> This is done through the use of the `gwcs` library.

Reviewer: ```suggestion
Even the `.ExtraCoords` class, which allows the user to add tabular data to axes, uses the `gwcs` library to store this information as a WCS.
This enables ndcube's coordinate transformation and plotting functions to leverage the same infrastructure, irrespective of whether the coordinates are functional or tabular.
```

   '2021-01-01T04:00:00.000' '2021-01-01T05:00:00.000'
   '2021-01-01T06:00:00.000' '2021-01-01T07:00:00.000'
   '2021-01-01T08:00:00.000' '2021-01-01T09:00:00.000']>

  >>> gwcs = TimeTableCoordinate(time_axis).wcs
  >>> gwcs
  <WCS(output_frame=TemporalFrame, input_frame=PixelFrame, forward_transform=Model: Tabular1D
  N_inputs: 1
  N_outputs: 1
  Parameters:
    points: (<Quantity [0., 1., 2., 3., 4., 5., 6., 7., 8., 9.] pix>,)
    lookup_table: [    0.  3600.  7200. 10800. 14400. 18000. 21600. 25200. 28800. 32400.] s
    method: linear
    fill_value: nan
    bounds_error: False)>

This `gwcs.WCS` object can then be passed to the constructor of `.NDCube` alongside your array and other parameters.

Combining Two Coordinates into a Single WCS
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

We can extend this example to be a space-space-time cube.
In this example we are going to utilize the ``mesh=`` keyword argument for the first time.
This keyword argument interprets the input to `.SkyCoordTableCoordinate` in a similar way to `numpy.meshgrid`.

.. code-block::

  >>> from astropy.coordinates import SkyCoord
  >>> from ndcube.extra_coords import SkyCoordTableCoordinate

>   >>> icrs_table = SkyCoord(range(10)*u.deg, range(10)*u.deg)

Reviewer: Suggest one of the axes in the `SkyCoord` have values of `range(10, 20)`*u.deg` so users can more easily see where the lat and lon values appear throughout the rest of the example.

.. _tabular_coordinates:

===================
Tabular Coordinates
===================

So far we have assumed that we are constructing an NDCube using a WCS that has been read from a file or explicitly created.
However, in some cases we may have tables giving the coordinate values at each pixel and converting these into a WCS manually can be tedious.
Therefore ndcube provides tools for constructing WCSes from such tables.

Tabular coordinates are useful when there is no mathematical description of the axis, or when it's a natural fit for the data you have.
It's worth considering that tabular coordinates are generally not as polished as a functional transform in a WCS.

> Therefore, if possiible, building a functional WCS for your coordinate system is highly recommended.

Reviewer: ```suggestion
Therefore, if possible, building a functional WCS for your coordinate system is highly recommended.
```

  >>> time_axis = Time("2021-01-01T00:00:00") + (list(range(10)) * u.hour)
  >>> time_axis
  <Time object: scale='utc' format='isot' value=['2021-01-01T00:00:00.000' '2021-01-01T01:00:00.000'
   '2021-01-01T02:00:00.000' '2021-01-01T03:00:00.000'
   '2021-01-01T04:00:00.000' '2021-01-01T05:00:00.000'
   '2021-01-01T06:00:00.000' '2021-01-01T07:00:00.000'
   '2021-01-01T08:00:00.000' '2021-01-01T09:00:00.000']>

  >>> gwcs = TimeTableCoordinate(time_axis).wcs
  >>> gwcs
  <WCS(output_frame=TemporalFrame, input_frame=PixelFrame, forward_transform=Model: Tabular1D
  N_inputs: 1
  N_outputs: 1
  Parameters:
    points: (<Quantity [0., 1., 2., 3., 4., 5., 6., 7., 8., 9.] pix>,)
    lookup_table: [    0.  3600.  7200. 10800. 14400. 18000. 21600. 25200. 28800. 32400.] s
    method: linear
    fill_value: nan
    bounds_error: False)>

This `gwcs.WCS` object can then be passed to the constructor of `.NDCube` alongside your array and other parameters.

Combining Two Coordinates into a Single WCS
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

We can extend this example to be a space-space-time cube.
In this example we are going to utilize the ``mesh=`` keyword argument for the first time.

> This keyword argument interprets the input to `.SkyCoordTableCoordinate` in a similar way to the way `numpy.meshgrid` works.

Reviewer: ```suggestion
This keyword argument interprets the input to `.SkyCoordTableCoordinate` in a similar way to `numpy.meshgrid`.
```

> When ``mesh=`` is used with `~astropy.coordinates.SkyCoord` the input is passed trough `numpy.meshgrid`, when it is used with `.QuantityTableCoordinate` the mesh is done dynamically.

Reviewer: ```suggestion
When ``mesh=`` is used with `~astropy.coordinates.SkyCoord` the input is passed through `numpy.meshgrid`, when it is used with `.QuantityTableCoordinate` the mesh is done dynamically.
```
